code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm

```diff
@@ -336,76 +336,139 @@
-	if(..())
-		var/area/A = get_area(src)
-		if(!A.outdoors && !(A in affected_targets))
-			for(var/atom/Stuff in A)
-				if(isturf(Stuff))
-					var/turf/T = Stuff
-					if((isspaceturf(T) || isfloorturf(T)) && NewTerrainFloors)
-						var/turf/open/O = T.ChangeTurf(NewTerrainFloors, flags = CHANGETURF_IGNORE_AIR)
-						if(prob(florachance) && NewFlora.len && !O.is_blocked_turf(TRUE))
-							var/atom/Picked = pick(NewFlora)
-							new Picked(O)
-						continue
-					if(iswallturf(T) && NewTerrainWalls)
-						T.ChangeTurf(NewTerrainWalls)
-						continue
-				if(istype(Stuff, /obj/structure/chair) && NewTerrainChairs)
-					var/obj/structure/chair/Original = Stuff
-					var/obj/structure/chair/C = new NewTerrainChairs(Original.loc)
-					C.setDir(Original.dir)
-					qdel(Stuff)
-					continue
-				if(istype(Stuff, /obj/structure/table) && NewTerrainTables)
-					new NewTerrainTables(Stuff.loc)
-					continue
-			affected_targets += A
+	. = ..()
+	if (!.)
+		return
+	var/area/current_area = get_area(src)
+	if (current_area in affected_targets)
+		return
+	if (terrain_theme.transform_area(current_area))
+		affected_targets += current_area
+
+/// Transforms an area's turfs and objects into a different theme
+/datum/crystal_warp_theme
+	/// Friendly name of theme
+	var/name = ""
+	/// Typepath of floor to replace open turfs with
+	var/floor
+	/// Typepath of wall to replace closed turfs with
+	var/wall
+	/// Typepath of object to replace chairs with
+	var/chair
+	/// Typepath of object to replace tables with
+	var/table
+	/// Typepath of things to potentially spawn on transformed open turfs
+	var/list/flora_and_fauna = list()
+	/// Chance per turf to create the things in the list above
+	var/flora_and_fauna_chance = 8
+
+/// Change appropriate objects in provided area to those matching our theme, and spawn some plants or animals
+/datum/crystal_warp_theme/proc/transform_area(area/target_area)
+	if (target_area.outdoors)
+		return FALSE
+	for(var/atom/thing in target_area)
+		if(isturf(thing))
+			replace_turf(thing)
+			continue
+		if(chair && istype(thing, /obj/structure/chair))
+			replace_object(thing, chair)
+			continue
+		if(table && istype(thing, /obj/structure/table))
+			replace_object(thing, table)
+			continue
+	return TRUE
+
+/// Replaces a turf with a different themed turf
+/datum/crystal_warp_theme/proc/replace_turf(turf/target_turf)
+	if(isindestructiblefloor(target_turf) || isindestructiblewall(target_turf))
+		return
+
+	if(iswallturf(target_turf))
+		if (wall)
+			target_turf.ChangeTurf(wall)
+		return
+
+	if(!isopenturf(target_turf))
+		return
+
+	if(length(flora_and_fauna) && !target_turf.is_blocked_turf(exclude_mobs = TRUE) && prob(flora_and_fauna_chance))
+		var/atom/new_flora_and_fauna = pick(flora_and_fauna)
+		new new_flora_and_fauna(target_turf)
+
+	if(isopenspaceturf(target_turf))
+		return // We replace space but not open space for whatever reason
```

Comment:
Doesn't this mean we'll place flora+fauna on openspace turfs? That seems silly, why not make this an early return at the beginning of the proc?